memory_allocator.c

```diff
@@ -0,0 +1,258 @@
+#include <common.h>
+#include <stdint.h>
+#include <stddef.h>
+#include "queue.h"
+#include "libkern.h"
+#include "include/memory_allocator.h"
+
+/*
+  TODO:
+  - use the mp_next field of malloc_pool
+*/
+
+#define MAGIC 0xC0DECAFE
+#define ALIGNMENT 8
+
+TAILQ_HEAD(mb_list, mem_block);
+TAILQ_HEAD(mp_arena, mem_arena);
+
+static inline int32_t abs(int32_t x)
+{
+  if (x < 0)
+    return -x;
+  return x;
+}
+
+typedef struct mem_block {
+  uint32_t mb_magic; /* if overwritten report a memory corruption error */
+  int32_t mb_size; /* size < 0 => free, size > 0 => alloc'd */
+  TAILQ_ENTRY(mem_block) mb_list;
```

Comment:
With union:
> 
> (gdb) print mb
> $11 = (mem_block_t *) 0x80001770
> 
> (gdb) print &mb->mb_list
> $10 = (struct {...} *) 0x80001778
> 
> (gdb) print mb->mb_data 
> $7 = 0x80001778
> 
> 
> 
> 
> 
<br>

Without union:
> 
> (gdb) print mb
> $9 = (mem_block_t *) 0x80001770
> 
> (gdb) print &mb->mb_list 
> $8 = (struct {...} *) 0x80001778
> 
> (gdb) print mb->mb_data
> $7 = 0x80001780

It does save us space!

However: [https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html](url)
> A structure containing a flexible array member, or a union containing such a structure (possibly recursively), may not be a member of a structure or an element of an array. (However, these uses are permitted by GCC as extensions.)

Do you still want to do this? I think that this feature is really shady. For example, this doesn't compile with my gcc version 4.8.4, even though that it, in my opinion, should:

> struct widget
> {
>   int a;
>   union {
>     struct {int x; int y;};
>     int b[];
>   };
> };
> 
> int main()
> {
>   return 0;
> }

> gcc -std=gnu99 file.c
> gcc -std=gnu11 file.c
> both give:
> untitled.c:6:9: error: flexible array member in union

It does compile with g++ though.